go/consensus/tendermint/light/p2p/p2p.go

```diff
@@ -0,0 +1,319 @@
+func NewLightClientProviderPool(ctx context.Context, chainContext string, chainID string, p2p rpc.P2P) LightClientProvidersPool {
+	pid := light.ProtocolID(chainContext)
+
+	mgr := rpc.NewPeerManager(p2p, pid)
+	rc := rpc.NewClient(p2p.Host(), pid)
+	rc.RegisterListener(mgr)
+
+	return &providersPool{
+		ctx:          ctx,
+		chainID:      chainID,
+		peerRegistry: make(map[core.PeerID]bool),
+		p2pMgr:       mgr,
+		rc:           rc,
+	}
+}
+
+type lightClientProvider struct {
+	sync.RWMutex
+
+	initOnce sync.Once
+	initCh   chan struct{}
+
+	logger *logging.Logger
+
+	chainID string
+
+	p2pMgr rpc.PeerManager
+	rc     rpc.Client
+
+	// Guarded by lock.
+	peerID            *core.PeerID
+	peerSelectionTime time.Time
+
+	pool *providersPool
+}
+
+func (lp *lightClientProvider) worker(ctx context.Context) {
+	ch, sub, err := lp.p2pMgr.WatchUpdates()
+	if err != nil {
+		lp.logger.Error("peer manager watch updates failure", "err", err)
+		return
+	}
+	defer sub.Close()
+
+	lp.refreshPeer()
+
+	for {
+		select {
+		case update := <-ch:
+			peerID := lp.getPeer()
+			switch {
+			case peerID == nil:
+				// If we have no peer and update was received try obtaining a peer.
+				lp.refreshPeer()
+			default:
+				// If we have a peer, ensure it was not removed.
+				if update.ID != *peerID {
+					continue
+				}
+				if update.PeerRemoved == nil {
+					continue
+				}
+				// Peer was removed, try obtaining a new peer.
+				lp.refreshPeer()
+			}
+		case <-ctx.Done():
+			lp.dropPeer()
+			return
+		}
+	}
+}
+
+func (lp *lightClientProvider) dropPeer() {
+	lp.Lock()
+	defer lp.Unlock()
+	lp.pool.Lock()
+	defer lp.pool.Unlock()
+	// Remove peer from pool.
+	if lp.peerID != nil {
+		delete(lp.pool.peerRegistry, *lp.peerID)
+	}
+	lp.peerID = nil
+}
+
+func (lp *lightClientProvider) getPeer() *core.PeerID {
+	lp.RLock()
+	defer lp.RUnlock()
+	return lp.peerID
+}
+
+func (lp *lightClientProvider) refreshPeer() {
+	lp.Lock()
+	defer lp.Unlock()
+	lp.pool.Lock()
+	defer lp.pool.Unlock()
+
+	lp.logger.Debug("refreshing peer", "peer", lp.peerID)
+
+	// Remove selected peer.
+	if lp.peerID != nil {
+		delete(lp.pool.peerRegistry, *lp.peerID)
+	}
+	lp.peerID = nil
+
+	// Select new peer.
+	peers := lp.p2pMgr.GetBestPeers()
+	for _, peerID := range peers {
+		if _, found := lp.pool.peerRegistry[peerID]; found {
+			continue
+		}
+		lp.logger.Debug("peer obtained", "peer_id", peerID)
+		p := peerID
+		lp.peerID = &p
+		lp.peerSelectionTime = time.Now()
+		lp.pool.peerRegistry[peerID] = true
+		lp.initOnce.Do(func() {
+			close(lp.initCh)
+		})
+		return
+	}
+	lp.logger.Debug("no p2p peer available")
+}
+
+// Initialized implements api.Provider.
+func (lp *lightClientProvider) Initialized() <-chan struct{} {
+	return lp.initCh
+}
+
+// MalevolentProvider implements api.Provider.
+func (lp *lightClientProvider) MalevolentProvider() {
+	lp.RLock()
+	defer lp.RUnlock()
+
+	// Ignore feedback right after peer was switched, as it could be a delayed
```

Comment:
The problem is that the peer can be switched for other reasons (light client exposes multiple methods and each could report a bad peer feedback which would then cause the peer to be switched internally). But using peer-feedback works fine because the feedback is reported for the actual underlying peer which served the response. 

For the light client, it's problematic because it is not aware of the underlying peer, it only knows that the provider misbehaved. Since we are already modifying our TM fork, we could instead make it aware of the peers. It would look something like this:
- the `MalevolentProvider` would be updated to: `MalevolentProvider(peer core.PeerID)` (or maybe even `string` type)
- the `LightBlock` & `ReportEvidence` methods would be updated to include the `peerID` in the response

The light client could then report the exact peer that provided the malicious response.

Reply:
> The light client could then report the exact peer that provided the malicious response.

Yeah I think that would be a better approach. I would use the `string` type for peer IDs to make it agnostic of what this is used for.

Reply:
Updated, see: https://github.com/oasisprotocol/tendermint/pull/192
Instead of modifying `LightBlock`, I added `LightBlockWithPeerID` to the Tendermint `Provider` interface, as `LightBlock` is used quite a bit throughout the code.

Also see updates in this PR.